### PR TITLE
Feature/rete 24 rsql

### PR DESCRIPTION
**Pull request by seanans**

### CODE REVIEW

File: src/main/java/com/laborete/LaboreteAPI/posts/controller/PostsController.java [hunk 1]
```diff
@@ -49,5 +49,11 @@ private ResponseEntity<PostDTO> getPostById(@PathVariable UUID id) {
     private List<PostDTO> filterPosts(@RequestBody FilterDTO searchValue) {
         return this.postsService.filterPosts(searchValue);
     }
+
+    @ApiOperation("Get Posts by RsqlFilter")
+    @PostMapping("/rsqlFilter")
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/posts/controller/PostsController.java]: why rsql? name it simply /filter

File: src/main/java/com/laborete/LaboreteAPI/posts/repository/PostRepository.java [hunk 2]
```diff
@@ -6,10 +6,13 @@
 import java.util.UUID;
 
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.stereotype.Repository;
 
 import java.util.UUID;
 
-public interface PostRepository extends JpaRepository<PostEntity, UUID> {
+@Repository
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/posts/repository/PostRepository.java]: It's not a correct usage of @Repository attribute. See spring's docs for more info.

File: src/main/java/com/laborete/LaboreteAPI/posts/service/PostsService.java [hunk 3]
```diff
@@ -16,5 +16,7 @@ public interface PostsService {
 
     List<PostDTO> filterPosts(FilterDTO filter);
 
+    List<PostDTO> rsqlFilterPosts(String rsqlFilter);
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/posts/service/PostsService.java]: `filterPosts`

File: src/main/java/com/laborete/LaboreteAPI/posts/service/PostsServiceImpl.java [hunk 4]
```diff
@@ -78,5 +81,17 @@ public List<PostDTO> filterPosts(FilterDTO filter) {
         List<PostEntity> postEntities = postRepository.findByTextContainingIgnoreCase(searchValue);
         return postMapper.postEntitiesListToPostDTOList(postEntities);
     }
+
+    @Override
+    public List<PostDTO> rsqlFilterPosts(String rsqlFilter) {
+        var parser = new RSQLParser();
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/posts/service/PostsServiceImpl.java]: I would make it a private field

File: src/main/java/com/laborete/LaboreteAPI/posts/service/PostsServiceImpl.java [hunk 5]
```diff
@@ -78,5 +81,17 @@ public List<PostDTO> filterPosts(FilterDTO filter) {
...
+        var root = parser.parse(rsqlFilter);
+
+        var spec = root.accept(new PostVisitor());
+        var postEntities = postRepository.findAll(spec);
+
+
+        return postEntities.stream().map(postMapper::postEntityToPostDTO).collect(Collectors.toList());
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/posts/service/PostsServiceImpl.java]: We could create a method that's able to map collections in the PostMapper

File: src/main/java/com/laborete/LaboreteAPI/profile/controller/UsersController.java [hunk 6]
```diff
@@ -39,6 +40,12 @@ private ResponseEntity<UserDTO> createUser(@RequestBody CreateUserDTO user) {
         );
     }
 
+    @ApiOperation(value = "Get users by filter")
+    @PostMapping("/rsqlFilter")
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/profile/controller/UsersController.java]: /filter

File: src/main/java/com/laborete/LaboreteAPI/profile/repository/UsersRepository.java [hunk 7]
```diff
@@ -2,12 +2,13 @@
 
 import com.laborete.LaboreteAPI.profile.entity.UserEntity;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 import org.springframework.stereotype.Repository;
 
 import java.util.Optional;
 import java.util.UUID;
 
 @Repository
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/profile/repository/UsersRepository.java]: !

File: src/main/java/com/laborete/LaboreteAPI/profile/services/UsersServiceImpl.java [hunk 8]
```diff
@@ -208,6 +213,17 @@ public ResponseEntity<HttpStatus> uploadUserBackground(String name, MultipartFil
         }
     }
 
+    @Override
+    public List<UserDTO> filterUsers(String rsqlFilter) {
+        var parser = new RSQLParser();
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/profile/services/UsersServiceImpl.java]: !

File: src/main/java/com/laborete/LaboreteAPI/rsql/UserVisitor.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+
+@Component
+public class UserVisitor implements RSQLVisitor<Specification<UserEntity>, Void> {
```
Comment (BobMakhlin) [inline src/main/java/com/laborete/LaboreteAPI/rsql/UserVisitor.java]: UserVisitor is a duplicate of PostVisitor. We should generalize the logic

**Review by BobMakhlin — APPROVED** [review]

File: src/main/java/com/laborete/LaboreteAPI/rsql/UserVisitor.java (hunk 9 above)
Comment (seanans) [reply src/main/java/com/laborete/LaboreteAPI/rsql/UserVisitor.java]: How?

File: src/main/java/com/laborete/LaboreteAPI/posts/service/PostsServiceImpl.java (hunk 4 above)
Comment (seanans) [reply src/main/java/com/laborete/LaboreteAPI/posts/service/PostsServiceImpl.java]: What the reason?

**Pull request merged**